src/image/hfe.c

```diff
@@ -375,11 +378,66 @@ static bool_t hfe_write_track(struct image *im)
             + (im->cur_track & 1) * 256
             + batch_off - im->hfe.write_batch.off
             + (off & 255);
-        for (i = 0; i < nr; i++)
+
+        i = 0;
+
+        if (im->hfe.fresh_seek && is_v3 && (off & 255) >= 1) {
+            /* Avoid writing in the middle of an opcode. */
+            char b = *(w-1);
+            if ((off & 255) >= 2)
+                if (*(w-2) == OP_SkipBits) {
+                    w++;
+                    i++;
+                }
+            if ((b & 0xf) == 0xf) {
+                switch (b) {
+                case OP_SkipBits:
+                    w += 2;
+                    i += 2;
+                    break;
+                case OP_Bitrate:
+                    w++;
+                    i++;
+                    break;
+                default:
+                    break;
+                }
+            }
+        }
+        im->hfe.fresh_seek = FALSE;
+
+        for (; i < nr; i++) {
+            if (is_v3 && (*w & 0xf) == 0xf) {
+                switch (*w) {
+                case OP_SkipBits:
```

Comment:
Keeping a skipbits opcode in the middle of your write? Doesn't that corrupt the write? I would expect overwrite the opcode, or turn the first two bytes into Nops.

Reply:
Maybe we could swap the bytes to nops, although then heads would become different lengths. At present I've only seen skipbits in places where it doesn't actually store data, like the end of the track. I've not seen a case where it matters, so I was choosing between keeping the old bits or dropping some new bits. It could potentially matter if used to align an OP_Rand, but I figure we'll deal with that when we find it. I didn't see it used to align OP_Index. I believe HxC doesn't support writes for HFEv3, so an option is to not support writing to a copy-protected sector.

The common case is OP_Bitrate, which HxCFloppyEmulator repeats often regardless of whether it is needed.

Reply:
I have an HFEv3 generator in Greaseweazle which will insert Skipbits opcodes to equalise the two sides of a cylinder. So I think skipping them is definitely bad, as there's nothing stopping this opcode appearing anywhere. Overwriting them is easy but does lengthen the track, I think it's the safer option though. Other options:
* Keep the opcode but change its data, putting you out of byte-sync with the bitcell ring. Morally superior solution but quite a pain in the arse?
* Overwrite with data. Remember how many bits ahead you now are. Fix up at the end with Nops and Skipbits as necessary. If you get too many whole bytes ahead meanwhile, insert some Nops. Also a PITA to implement but much less bit twiddling nonsense.

Reply:
> I have an HFEv3 generator in Greaseweazle which will insert Skipbits opcodes to equalise the two sides of a cylinder. So I think skipping them is definitely bad, as there's nothing stopping this opcode appearing anywhere.

Looks like Skipbits can only be at the end... of a chunk. It looks like [aligning the two sides is done with a NOP](https://github.com/keirf/greaseweazle/blob/56af717ad1d3286c0b2b40b6f7178f0dbd3c69ea/src/greaseweazle/image/hfe.py#L614C16-L614C20). So the problem isn't head alignment but bitcell size changes and weakness.

I hadn't expected to see the Bresenham-style error accumulation for the bitrate. That tells me a lot about its goals.

> Replace with 2 NOPs plus data. Does lengthen track by a few bits but not as bad as overwriting all three bytes.

I'll swap to this approach.

Reply:
> Looks like Skipbits can only be at the end... of a chunk. It looks like [aligning the two sides is done with a NOP](https://github.com/keirf/greaseweazle/blob/56af717ad1d3286c0b2b40b6f7178f0dbd3c69ea/src/greaseweazle/image/hfe.py#L614C16-L614C20). So the problem isn't head alignment but bitcell size changes and weakness.

Ah yes that's right. 

> I hadn't expected to see the Bresenham-style error accumulation for the bitrate. That tells me a lot about its goals.

The aim is maximum accuracy within a reasonable opcode budget. Tested with a lot of copy protected weird IPF images and works well. Raw dumps (with a lot of bitcell jitter) would need dejittering and clock recovery to discover the original bitrate-change boundaries; the existing code doesn't work well at all as-is.

> > Replace with 2 NOPs plus data. Does lengthen track by a few bits but not as bad as overwriting all three bytes.
> 
> I'll swap to this approach.

Good enough since as you note I'm only inserting this opcode at rate-change and weak-bit boundaries so it's not generally going to appear outside of copy protection tracks. So in practice this path will rarely trigger and we might need a synthetic image to test it.

Reply:
Things didn't work out quite as well as planned. I'm only able to replace one of the SkipBits bytes with a NOP, because the opcode might span flash blocks. Since it is expected to be rarer, figure that is still "good enough."